### ISSUE TITLE
Add `async` decoding attribute as default to the HTML image tags generated by WordPress

### ISSUE THREAD

**Issue by mitogh** [WordPress/performance#86]
We should evaluate the tradeoffs/benefits of adding the `decoding` attribute to images by default.

The idea would be to add the `decoding` attribute as `async` for nonvisible images similar to what happens with the `loading=lazy` attribute.

This ticket is created from core:

- https://core.trac.wordpress.org/ticket/53232

Useful links:

- https://developer.mozilla.org/en-US/docs/Web/API/HTMLImageElement/decoding

**Comment by dainemawer** [comment]
@mitogh sounds like this is a good idea! Do you think we should incorporate the code from the patch on the core trac or relook at things for this plugin?

I do feel like we need some kind of custom event in JS so that we know when an image has actually been decoded - I see they've been speaking about doing so on the WHATWG ticket: https://github.com/whatwg/html/issues/1920

The important benefit of adding this attribute is what was referenced here:
`Mac and iOS WebKit block painting on image decoding` - not sure about other operating systems.

The other caveat is images requested via CSS which we'll have no control over.

**Comment by mitogh** [comment]
I'm invision a patch from within this plugin that can eventually be merged into CORE. The idea would be similar to how the `lazy` attribute works for images applying the `decode=async` to images that are not the first image (or not inside of visible viewport). @dainemawer

**Comment by mehigh** [comment]
I've experienced using this even outside of AMP and have yet to encounter an edge-case where the decoding async had a negative consequence -> especially because the images in general loaded sooner when async. 

There is one particular caveat with this implementation - when this is a big image (not responsive, those utterly unoptimised types) above the fold with decoding async lacking width and height, it can introduce some layout shift. 
That seems like quite a peculiar edge case, given the fact that widths and heights are present  in a vast majority. 
The benefits far outweight the downsides here and applying it globally has a measurable influence on the interactivity of a page.

**Comment by mehigh** [comment]
I'll support the core ticket and provide before/after measurements, see if we can skip the need to put this in the plug-in and rather get that thing integrated as it is, given how trivial it is and the work already being completed.

**Comment by mehigh** [comment]
Have done the tests and covered the potential caveats I could think about, and this is passing with flying colors, as the lab+local tests have contradicted some of the conclusions I've drawn after reading the [specifications](https://html.spec.whatwg.org/multipage/embedded-content.html#dom-img-decoding).

Full result of the research posted on Trac :
https://core.trac.wordpress.org/ticket/53232#comment:34

TL;DR:
Explored several edge-cases in order to find potential downsides from including this attribute. Used webpagetest, GPSI & local lighthouse tests for all tests to confirm conclusions. Used performance profiling to go deeper and confirm the data/conclusions when the results were too consistent in the lab tests.
A big thumbs-up for this to be the default option for all images, not just the ones in the_content.

**Comment by mehigh** [comment]
Existing PR to core:
https://github.com/WordPress/wordpress-develop/pull/1853

**Comment by westonruter** [comment]
@mehigh Amazing and thorough research!

**Comment by westonruter** [comment]
I suppose the next question would be: should there be a way for site owners to selectively turn off `decoding=async` as they can for `loading=lazy`? In your research you found no downsides, then perhaps not?

**Comment by dainemawer** [comment]
It would also be good to note here that the presence of this attribute, is actually a browser hint, not something it has to commit to. It will effect the priority of tasks already set out by the browser vendor. Its still up to the browser to honour it. 

I know WP doesnt support SVG uploading due to security issues, but I thought it would be good to note here that the `decoding="async"` attribute can also be added to `<svg />` elements: https://developer.mozilla.org/en-US/docs/Web/API/SVGImageElement/decoding

**Comment by mehigh** [comment]
@westonruter 
I haven't found any benefits from selectively turning off decoding async. I think the one check that we could do (nice to have, not a must) is not output the decoding attribute a second time if it's already present, let's say if someone added it already to avoid bloating the HTML. 
Other than that, I din't think there's any need for a selective turn off.

Thanks for the detail @dainemawer . I've tested this only with linked SVG image files and it didn't create any negative consequences, it's great to hear the svg elements support it as well, as it confirms once again that adding this type of decoding on SVG files is OK to happen and that we don't have to filter these types of files out.

**Comment by adamsilverstein** [comment]
Let's continue to iterate on the core patch, it looks close to me!

**Comment by mitogh** [comment]
Just an FYI @bethanylang @eclarke1 that this ticket has been merged into core 6.1

- https://github.com/WordPress/wordpress-develop/commit/270259115c25e4f9b2143926d641dca43ce708ad

This means that we can close this one 🎉 

cc @peterwilsoncc @adamsilverstein @felixarntz

**Comment by mxbclang** [comment]
Yayyyy thank you for the heads up @mitogh and hooray for the merge! Closing this issue.

**Comment by peterwilsoncc** [comment]
Canonical commit is https://core.trac.wordpress.org/changeset/53480

**Comment by isaumya** [comment]
I am so happy to see this getting merged to WP core. I'm so happy that I opened the ticket for this in the first place. Great job guys. 🥳

**Comment by andrewstandoutreclame** [comment]
I've got a little error maybe due to this.
Uncaught SyntaxError: Unexpected token 'async' (at (index):1310:1290)
When json_encoding html (with a srcset url) the encoding="async" bit isn't getting encoded properly.
```
 <img decoding="async"  class=\"lazy \" data-src=\"
```
Ill dig some deeper and i'll follow it up.

**Comment by mxbclang** [comment]
Thanks, @andrewstandoutreclame! If you do continue to encounter this error, can you please <a href="https://github.com/WordPress/performance/issues/new?assignees=&labels=%5BType%5D+Bug&template=bug_report.md">open a new bug report issue</a> and refer back to this issue? That will help us triage this more efficiently. Thank you!

**Comment by isaumya** [comment]
Hi @andrewstandoutreclame,
That seems wired. Are you json encoding after the full HTML is generated cause then you are not supposed to have this issue.

**Comment by andrewstandoutreclame** [comment]
Thanks for the replies. I fixed it.

I could not find out how it put the decoding="async" in the HTML. The Wordpress filter changing the decoding attributes to sync didn't work. I checked for plugins but it didn't seem the case
However 'manually' putting in decoding="sync" on the image resolved the issue with the json_encode() function.


@isaumya I was generating the json via PHP (json_encode) right into a theme view (like page-contact.php)
```
<script>
        var people = <?php echo json_encode($all_people) ?>;
</script>
```

Where $all_people contains an object with HTML and an object with post data  like this: $person->fields, $person->html
